atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidatorTest.java

```diff
@@ -0,0 +1,294 @@
+    private static final String oldHostTwo = "old_host_two";
+    private static final String oldHostThree = "old_host_three";
+    private static final Set<String> oldHosts = ImmutableSet.of(oldHostOne, oldHostTwo, oldHostThree);
+    private static final Set<String> uuids = ImmutableSet.of("uuid1", "uuid2", "uuid3");
+
+    private static final HostIdResult DEFAULT_RESULT = HostIdResult.success(uuids);
+
+    private final CassandraTopologyValidationMetrics metrics =
+            CassandraTopologyValidationMetrics.of(new DefaultTaggedMetricRegistry());
+    private CassandraTopologyValidator validator = spy(new CassandraTopologyValidator(metrics));
+
+    @Test
+    public void retriesUntilNoNewHostsReturned() {
+        Iterator<String> uuidIterator = uuids.iterator();
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(newHosts);
+        EntryStream.of(allHosts)
+                .values()
+                .forEach(container -> setHostIds(
+                        Set.of(container),
+                        HostIdResult.success(Set.of(uuidIterator.next())),
+                        HostIdResult.success(uuids)));
+        assertThat(validator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        allHosts.keySet(),
+                        allHosts,
+                        HumanReadableDuration.milliseconds(1),
+                        HumanReadableDuration.seconds(20)))
+                .isEmpty();
+    }
+
+    @Test
+    public void retriesThrowsAndMarksFailureWhenNeverResolves() {
+        Iterator<String> uuidIterator = uuids.iterator();
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(newHosts);
+        EntryStream.of(allHosts)
+                .values()
+                .forEach(container -> setHostIds(Set.of(container), HostIdResult.success(Set.of(uuidIterator.next()))));
+        assertThat(validator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        allHosts.keySet(),
+                        allHosts,
+                        HumanReadableDuration.milliseconds(1),
+                        HumanReadableDuration.milliseconds(1)))
+                .isNotEmpty();
+        assertThat(metrics.validationFailures().getCount()).isEqualTo(1);
+        assertThat(metrics.validationLatency().getCount()).isEqualTo(1);
+    }
+
+    @Test
+    public void returnsEmptyWhenGivenNoServers() {
+        assertThat(validator.getNewHostsWithInconsistentTopologies(ImmutableSet.of(), ImmutableMap.of()))
+                .isEmpty();
+    }
+
+    @Test
+    public void throwsWhenAllHostsDoesNotContainNewHosts() {
+        assertThatThrownBy(() -> validator.getNewHostsWithInconsistentTopologies(
+                        ImmutableSet.of(createCassandraServer("foo")), ImmutableMap.of()))
+                .isInstanceOf(SafeIllegalArgumentException.class);
+    }
+
+    @Test
+    public void returnsEmptyWhenOnlyNewServersAndAllMatch() {
```

Comment:
Consider changing Match to ReportSameHostIds (I was confused given that this test also uses allHosts and allHosts (so new and all are the same host setup, which is actually irrelevant to the test)